File: src/formal/proof_cache.py
"""Disk-backed cache for verified Lean theorem results.

Cache key is a SHA-256 hash of what is being proved:
  - pure function source code, trailing whitespace removed
  - the property kind
  - the formal statement, with operator spelling and spacing normalised

Prose — the description, preconditions and assumptions — is deliberately not in
the key. See cache_key for why.

Only successful (verified) results are cached — failures are always retried.
"""

import hashlib
import json
import os
import re
from typing import TYPE_CHECKING

from .logger import get_logger, log
from .paths import PROOF_CACHE_DIR

if TYPE_CHECKING:
    from .property_verifier import PropertyResult

_log = get_logger(__name__)

_CACHE_DIR = PROOF_CACHE_DIR
_CACHE_TTL_DAYS = int(os.getenv("PROOF_CACHE_TTL_DAYS", "7"))


def _evict_expired() -> None:
    """Delete cache entries older than PROOF_CACHE_TTL_DAYS."""
    import time

    if not _CACHE_DIR.exists():
        return
    cutoff = time.time() - _CACHE_TTL_DAYS * 86400
    for path in _CACHE_DIR.glob("*.json"):
        if path.stat().st_mtime < cutoff:
            path.unlink(missing_ok=True)


def json_key(*parts: str) -> str:
    return hashlib.sha256("\n".join(parts).encode()).hexdigest()


def load_json(name: str) -> dict | None:
    path = _CACHE_DIR / f"{name}.json"
    if not path.exists():
        return None
    try:
        return json.loads(path.read_text())
    except Exception:
        return None


def save_json(name: str, payload: dict) -> None:
    """Best-effort — a cache miss must never be worse than a write failure."""
    try:
        _CACHE_DIR.mkdir(parents=True, exist_ok=True)
        (_CACHE_DIR / f"{name}.json").write_text(json.dumps(payload, indent=2))
    except OSError as e:
        log(_log, "CACHE", f"could not write {name[:20]}… — {e}")


_OPERATORS = {
    "∀": "forall",
    "∃": "exists",
    "→": "->",
    "⟶": "->",
    "↔": "<->",
    "∧": "/\\",
    "∨": "\\/",
    "¬": "not",
    "≠": "<>",
    "≤": "<=",
    "≥": ">=",
    "∈": " in ",
}


def normalise_formal(formal: str) -> str:
    """Reduce a formal statement to the form two writers of it should agree on.

    Operator spelling and spacing are free choices — `∀ x, p x → q x` and
    `forall x, p x -> q x` are one statement — and an agent picks differently
    from run to run where a fixed prompt at temperature 0 did not.
    """
    for symbol, ascii_form in _OPERATORS.items():
        formal = formal.replace(symbol, ascii_form)
    return re.sub(r"\s+", "", formal)


def normalise_code(function_code: str) -> str:
    """Indentation is meaning in Python, so only trailing and surrounding space goes."""
    return "\n".join(line.rstrip() for line in function_code.strip().splitlines())


def cache_key(function_code: str, kind: str, formal: str) -> str:
    """Identify a property by what is being proved, not by how it was described.

    The description, preconditions and assumptions that used to be mixed in here
    are English prose. A fixed prompt reproduces them verbatim, so the key worked
    while formal wrote them itself; an agent paraphrases, and every paraphrase was
    a fresh key and a re-proof. Across the 148 properties in a real run, the
    function, its kind and the normalised formal statement separate all of them.

    The prompt hash is gone with them. What is cached is a proof Lean accepted,
    and Lean's verdict does not depend on which prompt produced the theorem — and
    a prompt change that alters the formalisation changes `formal`, which changes
    the key anyway.
    """
    payload = "\n".join([normalise_code(function_code), kind.strip().lower(), normalise_formal(formal)])
    return hashlib.sha256(payload.encode()).hexdigest()


def load(key: str) -> "PropertyResult | None":
    from .property_verifier import PropertyResult

    path = _CACHE_DIR / f"{key}.json"
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text())
        return PropertyResult(**data)
    except Exception:
        return None


def save(key: str, result: "PropertyResult") -> None:
    """Best-effort — the cache is an optimisation and never changes a verdict."""
    try:
        _CACHE_DIR.mkdir(parents=True, exist_ok=True)
        path = _CACHE_DIR / f"{key}.json"
        path.write_text(json.dumps(result.__dict__, indent=2))
        _evict_expired()
    except OSError as e:
        log(_log, "CACHE", f"could not write {key[:12]}… — {e}")

File: rust/formal-core/src/proof_cache.rs
//! Identification of what is being proved.
//!
//! The key is a SHA-256 hash of:
//!   - pure function source code, trailing whitespace removed
//!   - the property kind
//!   - the formal statement, with operator spelling and spacing normalised
//!
//! Prose — the description, preconditions and assumptions — is deliberately not
//! in the key. See [`cache_key`] for why.

use std::sync::OnceLock;

use regex::{
    Captures,
    Regex,
};
use sha2::{
    Digest,
    Sha256,
};

use crate::pystr;

/// Canonical form is the symbol, not the word. Words were the wrong direction:
/// with whitespace stripped, `∀x` became `forallx`, which is also what the
/// identifier `forallx` becomes. The symbols cannot occur inside an identifier,
/// so they can. Longest first — `<->` contains `->`.
const ASCII_OPERATORS: &[(&str, &str)] = &[
    ("<->", "↔"),
    ("->", "→"),
    ("/\\", "∧"),
    ("\\/", "∨"),
    ("<>", "≠"),
    ("<=", "≤"),
    (">=", "≥"),
    ("⟶", "→"),
];

/// Spelled as words, so they only count on a word boundary: `in` inside `ainb`
/// is not the membership operator, and treating it as one merged unrelated
/// statements.
const WORD_OPERATORS: &[(&str, &str)] =
    &[("forall", "∀"), ("exists", "∃"), ("not", "¬"), ("in", "∈")];

fn word_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| {
        let alternation = WORD_OPERATORS
            .iter()
            .map(|&(word, _)| word)
            .collect::<Vec<_>>()
            .join("|");
        Regex::new(&format!(r"\b({alternation})\b")).expect("the operator words are literals")
    })
}

/// Reduce a formal statement to the form two writers of it should agree on.
///
/// Operator spelling and spacing are free choices — `∀ x, p x → q x` and
/// `forall x, p x -> q x` are one statement — and an agent picks differently
/// from run to run where a fixed prompt at temperature 0 did not.
///
/// Word-spelled operators are matched on word boundaries, and only before the
/// whitespace goes. Replacing them afterwards, or by substring, merges
/// statements that merely contain the letters: `a∈b` and the unrelated `ainb`
/// both reduced to `ainb` under the previous version.
#[must_use]
pub fn normalise_formal(formal: &str) -> String {
    let mut formal = formal.to_string();
    for &(ascii_form, symbol) in ASCII_OPERATORS {
        formal = formal.replace(ascii_form, symbol);
    }
    let formal = word_pattern().replace_all(&formal, |caps: &Captures| {
        let word = &caps[1];
        WORD_OPERATORS
            .iter()
            .find(|&&(candidate, _)| candidate == word)
            .map_or_else(|| word.to_string(), |&(_, symbol)| symbol.to_string())
    });
    formal.chars().filter(|&c| !pystr::is_space(c)).collect()
}

/// Indentation is meaning in Python, so only trailing and surrounding space goes.
#[must_use]
pub fn normalise_code(function_code: &str) -> String {
    pystr::splitlines(pystr::strip(function_code))
        .into_iter()
        .map(pystr::rstrip)
        .collect::<Vec<_>>()
        .join("\n")
}

/// Join fields so that no field can imitate the boundary between two others.
///
/// Joining on a newline was ambiguous: normalised code and the kind may both
/// contain one, so `("X\na", "b", "c")` and `("X", "a\nb", "c")` produced the
/// same payload and therefore the same key — two distinct properties sharing
/// one cached proof. Length-prefixing each field removes the ambiguity whatever
/// the field contains.
///
/// The length is a count of characters, not of bytes. Normalisation puts
/// multi-byte operators into the statement, so the two differ for most real
/// inputs and a byte count moves every key.
fn framed(parts: &[&str]) -> String {
    let mut payload = String::new();
    for part in parts {
        payload.push_str(&part.chars().count().to_string());
        payload.push(':');
        payload.push_str(part);
    }
    payload
}

/// Identify a property by what is being proved, not by how it was described.
///
/// The description, preconditions and assumptions that used to be mixed in here
/// are English prose. A fixed prompt reproduces them verbatim, so the key worked
/// while formal wrote them itself; an agent paraphrases, and every paraphrase
/// was a fresh key and a re-proof. Across the 148 properties observed in a real
/// run, the function, its kind and the normalised formal statement separated all
/// of them — an observation about that corpus, not a theorem about the function.
/// Distinctness of digests would rest on sha256, which nothing here establishes.
///
/// The prompt hash is gone with them. What is cached is a proof Lean accepted,
/// and Lean's verdict does not depend on which prompt produced the theorem — and
/// a prompt change that alters the formalisation changes `formal`, which changes
/// the key anyway.
#[must_use]
pub fn cache_key(function_code: &str, kind: &str, formal: &str) -> String {
    let payload = framed(&[
        &normalise_code(function_code),
        &pystr::strip(kind).to_lowercase(),
        &normalise_formal(formal),
    ]);
    format!("{:x}", Sha256::digest(payload.as_bytes()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn iff_is_matched_before_the_arrow_inside_it() {
        assert_eq!(normalise_formal("a <-> b"), "a↔b");
    }

    #[test]
    fn a_word_operator_is_not_matched_inside_an_identifier() {
        assert_ne!(normalise_formal("a in b"), normalise_formal("ainb"));
        assert_eq!(normalise_formal("a in b"), "a∈b");
        assert_eq!(normalise_formal("ainb"), "ainb");
    }

    #[test]
    fn normalising_twice_changes_nothing() {
        let once = normalise_formal("forall x, p x -> q x");
        assert_eq!(normalise_formal(&once), once);
    }

    #[test]
    fn indentation_survives_but_trailing_space_does_not() {
        assert_eq!(
            normalise_code("\ndef f():\n    return 1   \n\n"),
            "def f():\n    return 1"
        );
    }

    #[test]
    fn no_field_can_imitate_the_boundary_of_another() {
        assert_ne!(cache_key("X\na", "b", "c"), cache_key("X", "a\nb", "c"));
    }

    #[test]
    fn the_kind_is_matched_case_insensitively() {
        assert_eq!(
            cache_key("f", "  Invariant  ", "p"),
            cache_key("f", "invariant", "p")
        );
    }

    #[test]
    fn the_frame_counts_characters_and_not_bytes() {
        assert_eq!(framed(&["∀"]), "1:∀");
    }
}
